listreceivedbyaddress Filter Address

NicolasDorier opened the pull request:
Supersede https://github.com/bitcoin/bitcoin/pull/9503 created by @JeremyRubin , I will maintain it.

JeremyRubin committed:
Add address filtering to listreceivedbyaddress

JeremyRubin committed:
Add tests of listreceivedbyaddress address filtering

NicolasDorier commented:
1. Rebased and fixed conflicts
2. Addressed @kallewoof nits
3. Add the new parameter to `client.cpp`

kallewoof reviewed (approved):
utACK 8c373d5feedb428d234278d484e5d2a932010fae

NicolasDorier commented:
@luke-jr @jnewbery nit addressed, can you re-ACK?

jnewbery commented on qa/rpc-tests/receivedby.py:
```diff
@@ -62,10 +62,22 @@ def run_test(self):
+        #Test Address filtering
+        #Only on addr
+        expected = {"address":addr, "account":"", "amount":Decimal("0.1"), "confirmations":10, "txids":[txid,]}
+        res = self.nodes[1].listreceivedbyaddress(0, True, True, addr)
```
Nit: Consider using named arguments here, instead of positional arguments (since the meaning of the arguments is not clear without having to go back to the definition of listreceivedbyaddress).

NicolasDorier replied:
Listreceivedbyaddress is not known at compile time, and thus has no definition in python. This is interpreted at runtime. So I can't really use named argument. Would inline comment be good ?

jnewbery replied:
You can use named arguments in the RPC calls. See https://github.com/bitcoin/bitcoin/pull/9983 for example.

jnewbery commented on qa/rpc-tests/receivedby.py:
```diff
@@ -62,10 +62,22 @@ def run_test(self):
+                           {"address":empty_addr},
+                           {"address":empty_addr, "account":"", "amount":0, "confirmations":0, "txids":[]})
+
+        #Test Address filtering
```
I'd prefer this test if `other_addr` (currently declared on line 78) received some funds before you did your tests on `listreceivedbyaddress()` with a filter address. This isn't currently testing the case where the wallet contains multiple addresses with funds and you want to see the transactions received by one of those addresses. Ideally the test case would be:

- define addr and send funds to it (already done for you in lines 42-43)
- define other_addr and send funds to it
- call `listreceivedbyaddress()` filtering on `addr`. Assert only one transaction is returned.
- call `listreceivedbyaddress()` filtering on `other_addr`. Assert only one transaction is returned.
- call `listreceivedbyaddress()` with no filtering. Assert both transactions are returned.

jnewbery commented:
Looks good. A couple of nits and a suggestion on improving the test case.

JeremyRubin commented:
utAck

NicolasDorier commented:
@jnewbery I improved the tests, and fixed the alignement. I can't use named parameter in the test though.

NicolasDorier commented:
added named parameters to one call to listreceivedbyaddress  in the tests for better readability.

jnewbery commented:
line 73 `minConf` needs to be `minconf`.

with that change, tested ACK.

NicolasDorier commented:
@jnewbery thanks done ;)

EthanHeilman commented:
utACK - the code looks good and this is very useful functionality.

NicolasDorier commented:
@TheBlueMatt I fixed nits and added tests, can you re-ACK ?

NicolasDorier commented:
https://github.com/bitcoin/bitcoin/pull/7729 should also work for TB purpose.

nopara73 commented:
ACK - works fine.  
This could be (not sure if should be) combined with fixing this issue for correct GUI experience: https://github.com/bitcoin/bitcoin/issues/9192

NicolasDorier commented:
Closing. I am using a workaround now, I caching the whole `listtransaction` at every block. This call would have been super useful though.

jnewbery commented:
:disappointed: I'm happy to reACK if you decide to reopen this!

NicolasDorier commented:
@jnewbery, if you have interest, I am reopening, I guess it does not hurt to keep it opens, it is not high maintenance PR. I was thinking to allow multiple addresses to be passed in the filter, instead of only one.

For NTumbleBit, I changed the design, I have a method called `GetTransactions(address)` which should send back the received transactions as well as the outputs one of this address. `listreceivedbyaddress` would only least what the address received.

The way I ended up implemeting my need is by caching `listtransaction` then for each transaction in it, caching `gettransaction` in a internal database. I now have good performance, and I am not sure this RPC Method would improve my case.

sipa commented:
@NicolasDorier Do you actually need the wallet functionality (balance tracking, UTXO management, coin selection, ...), or just a means to be notified of certain transactions?

nopara73 commented:
@sipa Need the wallet functionality for NTumbleBit.

sipa commented:
Can you clarify what functionality it relies on?

nopara73 commented:
A lot, but I am not sure which are relevant to this discussion with the workaround. @NicolasDorier will have a better insight. https://github.com/NTumbleBit/NTumbleBit/tree/master/NTumbleBit.TumblerServer/Services/RPCServices

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1206,6 +1206,17 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
 
+    bool fFilterAddress = false;
+    CTxDestination filterAddress =  CNoDestination();
+    if (!fByAccounts && params.size() > 3) {
```
Use a simpler check:
```cpp
CBitcoinAddress address(request.params[3].get_str());
    if (!address.IsValid())
```

NicolasDorier replied:
This does not apply anymore.

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1240,10 +1255,24 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
+
+    // Create mapAddressBook iterator
+    // If we aren't filtering, go from begin() to end()
+    auto start = pwallet->mapAddressBook.begin();
```
This loop-or-find logic seems to be a bit wired.
I probably would have factored out the JSON object packing and either called the new function from within the range based loop or after the `find()`.

jonasschnelli commented:
Needs rebase.

Concept ACK. I think the by address filter is in general useful.

NicolasDorier commented:
@sipa just saw now your previous question.
As far as TB is concerned, I am using balance tracking, UTXO management, and coin selection management of Bitcoin Core for funding the escrow transactions.

I am also using Core as a block explorer: There is some well known address on which I would like to be notified is money is sent *to* it, or sent *from*. The method allows me to know when something is sent to the address.
However, I still need a new rpc method for knowing if something is sent *from* an address.

Until I have the two features, my only way is to poll periodically `listtransaction` then fetch all transaction with `gettransaction`, and see if the relevant addresses are inside. This work well but is very cumbersome.

luke-jr commented on test/functional/receivedby.py:
```diff
@@ -62,10 +62,42 @@ def run_test(self):
+        assert_equal(len(res), 1)
+        #Another address receive money
+        res = self.nodes[1].listreceivedbyaddress(0, True, True)
+        assert_equal(len(res), 3) #Right now 3 entries
```
Note, this needs to be 2->3 (instead of 3->4) due to removal of the "default address".

luke-jr commented:
Rebased here: https://github.com/luke-jr/bitcoin/tree/listreceivedbyaddress-filtered

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1206,6 +1206,17 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
+    if (!fByAccounts && params.size() > 3) {
+        filterAddress =  CBitcoinAddress(params[3].get_str()).Get();
+        CTxDestination nulladdress = CNoDestination();
+        if (filterAddress == nulladdress) {
```
Just `if (filterAddress == CNoDestination())`?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1206,6 +1206,17 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
+    bool fFilterAddress = false;
+    CTxDestination filterAddress =  CNoDestination();
+    if (!fByAccounts && params.size() > 3) {
+        filterAddress =  CBitcoinAddress(params[3].get_str()).Get();
```
Remove extra space.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1206,6 +1206,17 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool fFilterAddress = false;
+    CTxDestination filterAddress =  CNoDestination();
```
Remove extra space.

promag commented on src/rpc/client.cpp:
```diff
@@ -44,6 +44,7 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "listreceivedbyaddress", 0, "minconf" },
     { "listreceivedbyaddress", 1, "include_empty" },
     { "listreceivedbyaddress", 2, "include_watchonly" },
+    { "listreceivedbyaddress", 3, "only_address" },
```
Not sure what others think, but I kind of prefer `filter_address`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1206,6 +1206,17 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
+        filterAddress =  CBitcoinAddress(params[3].get_str()).Get();
+        CTxDestination nulladdress = CNoDestination();
+        if (filterAddress == nulladdress) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "only_address parameter was invalid");
```
Missing test `assert_raises_rpc_error`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1314,15 +1343,15 @@ UniValue listreceivedbyaddress(const JSONRPCRequest& request)
             "2. include_empty     (bool, optional, default=false) Whether to include addresses that haven't received any payments.\n"
             "3. include_watchonly (bool, optional, default=false) Whether to include watch-only addresses (see 'importaddress').\n"
-
+            "4. only_address      (string, optional) If present, only return information on this address. Otherwise, return all information.\n"
```
Drop `Otherwise, return all information`?

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1240,10 +1255,24 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
+        }
+    }
+
+    for(auto item_it = start; item_it != end; ++item_it)
```
Nit, space after `for`, `{` on this line.

promag reviewed (commented):
Some minor comments, new code could use current convention, for instance, s/filterAddress/filter_address.

NicolasDorier commented:
Closing this I workedaround the need of this

JeremyRubin commented:
@NicolasDorier I think this should reopen, AFAIK LN people would like this too... @cfromknecht @Roasbeef

NicolasDorier commented:
ok rebasing

NicolasDorier commented:
Rebased, fixed the nits of @luke-jr and @promag .

I renamed `only_address` to `address_filter` so that if later we decide it is also possible to pass an array of address instead of a single address, we don't need to rename anything. It is more generic.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1403,6 +1403,16 @@ UniValue ListReceived(CWallet * const pwallet, const UniValue& params, bool fByA
         if(params[2].get_bool())
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool fFilterAddress = false;
```
The code convention has changed a bit recently. Now, we don't put the type as prefix, and we try to use snake case. I would rename this `has_filter_address` and the below to `filter_address`. You can leave existing vars as is, if you don't want the diff to grow for no reason though (like `fByAccounts` below).

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3837,7 +3866,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "listlockunspent",                  &listlockunspent,               {} },
     { "wallet",             "listreceivedbyaccount",            &listreceivedbyaccount,         {"minconf","include_empty","include_watchonly"} },
-    { "wallet",             "listreceivedbyaddress",            &listreceivedbyaddress,         {"minconf","include_empty","include_watchonly"} },
+    { "wallet",             "listreceivedbyaddress",            &listreceivedbyaddress,         {"minconf","include_empty","include_watchonly","address_filter" } },
```
No space after `"` (end of line) -- compare to other lines.

kallewoof reviewed (approved):
utACK

NicolasDorier commented:
Addressed @kallewoof nits

kallewoof reviewed (approved):
utACK f087613719026bcd5cba95ec64c19361fcc71ecf

laanwj commented:
utACK f08761371

luke-jr commented on test/functional/wallet_listreceivedby.py:
```diff
@@ -45,10 +45,44 @@ def run_test(self):
+                            {"address": empty_addr},
+                            {"address": empty_addr, "account": "", "amount": 0, "confirmations": 0, "txids": []})
+
+        #Test Address filtering
```
Space after `#` is expected now it seems.

luke-jr commented on test/functional/wallet_listreceivedby.py:
```diff
@@ -45,10 +45,44 @@ def run_test(self):
+
+        #Test Address filtering
+        #Only on addr
+        expected = {"address":addr, "account":"", "amount":Decimal("0.1"), "confirmations":10, "txids":[txid,]}
```
Space after `:` is expected now it seems.
